File: 2021/09/27/dominant_index.py
"""
https://leetcode.com/problems/largest-number-at-least-twice-of-others/
Author: Andrew Jarombek
Date: 9/27/2021
"""

from typing import List


def dominant_index(nums: List[int]) -> int:
    if len(nums) == 1:
        return 0

    second_largest = 0
    largest = 0
    largest_index = 0

    for index, num in enumerate(nums):
        if num >= largest:
            largest_index = index
            second_largest = largest
            largest = num
        elif num > second_largest:
            second_largest = num

    return largest_index if largest >= second_largest * 2 else -1

File: 2022/12/27/dominantIndex.js
/**
 * https://leetcode.com/problems/largest-number-at-least-twice-of-others/
 * @author Andrew Jarombek
 * @since 12/27/2022
 */

const dominantIndex = function(nums) {
    const values = [nums[0], -1];
    let index = 0;

    for (let i = 1; i < nums.length; i++) {
        const num = nums[i];

        if (num > values[0]) {
            values[1] = values[0];
            values[0] = num;
            index = i;
        } else if (num > values[1]) {
            values[1] = num;
        }
    }

    return (values[0] >= (values[1] * 2)) ? index : -1;
}

console.assert(dominantIndex([3,6,1,0]) === 1);
console.assert(dominantIndex([1,2,3,4]) === -1);
